### DISCUSSION TITLE
Deferring dynamic import to client on Next 13 build

### DISCUSSION THREAD

**Discussion by cereallarceny** [vercel/next.js#44940] [Help]
Hello!

I've integrated [TinaCMS](https://tina.io/) into my Next 13 app, using the experimental app directory. As part of this, my CMS schema allows for the selection of an icon from the popular [react-icons](https://react-icons.github.io/react-icons/) library. This allows content authors to be able to customize the icon for any given page section to their liking, without developer intervention. The cost of this decision is that we now have pretty sizeable "First Load JS" (upwards of 700-800kb per page). I temporarily disabled dynamic icons and re-ran the build, and immediately noticed a roughly ~400kb bundle size drop - effectively a 50-60% improvement on some pages.

With that said, I can't realistically disable this functionality, but I'd love to see the loading of icons deferred to the client and off of my "First Load JS". On the other hand, it's somewhat hard for me to tell if that "First Load JS" includes client-side components. Since Next 13 dropped, I've been finding it hard to determine what in my component stack is being rendered as a client component and what's being rendered as a server component. Anyhow, let's dive into the code I've got so far, maybe there's a way to bring these build sizes down:

```tsx
'use client';

import dynamic from 'next/dynamic';
import { Icon, IconProps } from 'ui';

const iconLoader = (icon) =>
  dynamic(() => import('react-icons/fa').then((mod) => mod[icon]), {
    ssr: false,
  });

const DynamicIcon = ({ icon, ...props }: { icon: string } & IconProps) => (
  <Icon as={iconLoader(icon)} {...props} />
);

export default DynamicIcon;
```

This allows the implementation to look something like this: `<DynamicIcon icon="FaBeer" {...someOtherProps} />`. My CMS is the one providing me the string `"FaBeer"` in this case.

I'll be totally honest, the use of `dynamic()`, `ssr: false`, `'use client'` are just me throwing things at the wall. I'm not committed to any of those decisions and will take any help I can get in reducing the bundle size. To cap it off, here's a screenshot of my editor yelling at me for such a massive, shameful import. 😆 

<img width="891" alt="Screenshot 2023-01-16 at 5 44 49 PM" src="https://user-images.githubusercontent.com/1297930/212781171-9e4d6c02-ae7a-4d43-818b-9c9072254c66.png">

**Comment by icyJoseph** [comment]
Hide the Icon behind a `React.useState` flag, which is only set to true within a `useEffect`. That way, the bundle will only be require from the client. 

In Server Components, and React 18, suspense boundaries might be awaited on the server, depending on whether or not we tell `react-dom/server` to await for them. I do believe Next.js waits for all async work to flush in SSR, so your icons end up bundled too. A safe bet is to render as much as possible, even with ARIA fallback, and load the icons client side after the first browser paint, AKA, useEffect with no dependencies.

**Reply by cereallarceny** [reply]
@icyJoseph Do you mean something like this?

```tsx
'use client';

import dynamic from 'next/dynamic';
import { useEffect } from 'react';
import { useState } from 'react';
import { Icon, IconProps } from 'ui';

const DynamicIcon = ({ icon, ...props }: { icon: string } & IconProps) => {
  const [isVisible, setIsVisible] = useState(false);

  useEffect(() => {
    setIsVisible(true);
  }, []);

  if (!isVisible) return null;

  const iconLoader = (icon) =>
    dynamic(() => import('react-icons/fa').then((mod) => mod[icon]));

  return <Icon as={iconLoader(icon)} {...props} />;
};

export default DynamicIcon;
```

I still get the same build size as before.

**Reply by icyJoseph** [reply]
What's up with `ui`? Does that import "react-icons/fa" itself?

I imagine you maybe import a type or something from "react-icons/fa" inside of "ui"?

**Reply by cereallarceny** [reply]
The ui package wraps around Chakra UI since that library doesn't support server components yet. I'm just exporting the Icon component and IconProps type.

Either way, what I posted doesn't seem to affect bundle size. Did I follow your suggestion incorrectly or are we dealing with a deeper problem?

**Reply by icyJoseph** [reply]
Well, I reckon the issue is that react-icons/fa might be imported by other module, which means that webpack will bundle it as part of the main bundle, in spite of any dynamic loading.

Oh right, one thing though:

```tsx
'use client';

import dynamic from 'next/dynamic';
import { useEffect, useState } from 'react';
import { Icon, IconProps } from 'ui';

const lazyIcons =  dynamic(() => import('react-icons/fa'))

const DynamicIcon = ({ icon, ...props }: { icon: string } & IconProps) => {
  const [isVisible, setIsVisible] = useState(false);

  useEffect(() => {
    setIsVisible(true);
  }, []);

  if (!isVisible) return null;

  const iconLoader = (icon) => lazyIcons.then((mod) => mod[icon]));

  return <Icon as={iconLoader(icon)} {...props} />;
};

export default DynamicIcon;
```

Try with something like that, otherwise, you could load the icons in the useEffect:

```tsx
'use client';

import dynamic from 'next/dynamic';
import { useEffect, useState } from 'react';
import { Icon, IconProps } from 'ui';

const DynamicIcon = ({ icon, ...props }: { icon: string } & IconProps) => {
  const [icons, setIcons] = useState<typeof import("react-icons/fa") 
   null>(null);

  useEffect(() => {
        dynamic(() => import('react-icons/fa').then((mod) =>  setIcons(mod)));
  }, []);

  if (!icons) return null;


  return <Icon as={icons[icon]} {...props} />;
};

export default DynamicIcon;
```

Warning the above is pseudocode...

**Reply by cereallarceny** [reply]
@icyJoseph Neither of these approaches (when corrected from pseudocode) had any effect over the build size. However, I can confirm that this component alone IS responsible for the change, not react-icons itself. I'm using named exports of react-icons elsewhere, which appropriately has only a menial effect on build size. This `DynamicIcon` component is the one having the outsized effect on bundle size. So I'm convinced we're in the right place, but none of these solutions seem to be doing the trick.

I do wonder if perhaps we're thinking about this in the wrong way. Maybe that First Load JS is accounting for client-side components. Technically speaking, since all of the components that use this component are pre-rendered, I should be able to compute them in advance at build time. In that scenario, I suppose they don't HAVE to be client-side components, right? Since Next 13 was released, I've found it very difficult in some scenarios to determine how components are interpreted by the NextJS compiler. Maybe instead of forcing this to be a client component, we should force it to be a server component?

It really depends on how Next 13 is calculating "First Load JS". That's less clear to me now that server components are a thing.

**Reply by icyJoseph** [reply]
Alright, I've now read a bit more carefully your comments, and tried out a clean install of create-next-app with react-icons/fa, and I don't really see a 400kB increase on size when using them. 

Quite simply, if a component is a server component is pre-rendered Server Side and no JS is shipped for it ~ so to speak, but if you have client components, then a runtime is loaded, once, for all of the client components, plus their dependencies. 

When I run bundle analyzer on my toy app, I see that the client side contains about 8kB of react-icons/fa, whereas the server (node.js) side bundle has 400kB ~ ish. That's just bringing in one icon though:

![Screenshot 2023-01-18 at 09 18 26](https://user-images.githubusercontent.com/21013447/213119332-88c2c796-cc7e-4e0c-98aa-f41b9b900920.png)

```tsx
"use client";

import { FaBeer } from "react-icons/fa";

export const ClientComponent = () => {
  return (
    <div>
      <p>Beer:</p>
      <p>
        <FaBeer />
      </p>
    </div>
  );
};
```

Now let's do a dynamic import of the whole thing:

```tsx
"use client";

import { useEffect, useState } from "react";

export const ClientComponent = () => {
  const [icons, setIcons] = useState<typeof import("react-icons/fa") | null>(
    null
  );

  useEffect(() => {
    import("react-icons/fa").then((mod) => setIcons(mod)); // import-cost now complains about this!
  }, []);

  if (!icons) return null;

  const Icon = icons["FaBeer"];
  return (
    <div>
      <p>Beer:</p>

      <p>
        <Icon />
      </p>
    </div>
  );
};

```

And when running the bundle analyzer I now see a huge chunk of 419.82 kB, which is the entire icon library.

![Screenshot 2023-01-18 at 09 20 26](https://user-images.githubusercontent.com/21013447/213119696-a24de5b9-97b2-436d-aec6-e04f9f024d54.png)

The Chrome DevTools for coverage benchmark, indicate a 1.3 MB file, with 98% unused Javascript has been downloaded, and yes, it is the Icon library!

![Screenshot 2023-01-18 at 09 19 48](https://user-images.githubusercontent.com/21013447/213119524-74a04a7b-bae7-4642-8ca1-68a6adfaee0c.png)

Interestingly, this is a separate chunk, with only icons.

Let's try the same in pages only, and we'll see the same issue! (Notice the hash name of the icon bundle is the same, but I've now moved the "client component" as a good ol' page)

![Screenshot 2023-01-18 at 09 21 52](https://user-images.githubusercontent.com/21013447/213119956-6d5a28a3-b87b-46bc-9512-e66d0ca8e21d.png)

So now we know the Server Components, appDir has nothing to do with it. There's a fundamental issue with what we are trying to achieve.

```tsx
import { useEffect, useState } from "react";

const IconPage = () => {
  const [icons, setIcons] = useState<typeof import("react-icons/fa") | null>(
    null
  );

  useEffect(() => {
    import("react-icons/fa").then((mod) => setIcons(mod));
  }, []);

  if (!icons) return null;

  const Icon = icons["FaBeer"];
  return (
    <div>
      <p>Beer:</p>

      <p>
        <Icon />
      </p>
    </div>
  );
};

export default IconPage;
```

The root issue is that when we create a dynamic import, we bundle up the entire module into a single file. When we do `import('react-icons/fa')` we are throwing away the entire module separation they've made for us, and we are telling Webpack to bundle it all up into one chunk. 

What we want is to make a request per icon chunk! One icon - one chunk.

## A Hack

We can abuse dynamic imports, by breaking a rule, we will not have statically defined dynamic imports, we'll let it be a wild card, this has implications that you'll have to deal with yourself.

First you must install: `@react-icons/all-files`

> Warning: Bundle analyzer timing will sky rocket with this hack

```tsx
import { useEffect, useState } from "react";

import type { IconType } from "@react-icons/all-files";

const IconPage = ({ iconName = "FaBeer" }: { iconName: string }) => {
  const [icon, setIcon] = useState<
    { [key in typeof iconName]: IconType } | null
  >(null);

  useEffect(() => {
    import(`@react-icons/all-files/fa/${iconName}.js`).then((mod) =>
      setIcon(mod)
    );
  }, [iconName]);

  if (!icon) return null;

  const Icon = icon[iconName];

  return (
    <div>
      <p>Beer:</p>

      <p>
        <Icon />
      </p>
    </div>
  );
};

export default IconPage;

```

And now if we build... get ready for this...

![Screenshot 2023-01-18 at 09 52 15](https://user-images.githubusercontent.com/21013447/213126276-7df5fad7-caea-44d3-95c9-d20f64da7d64.png)

Every little box in this screenshot is an icon!

Now I check the Chrome DevTools:

![Screenshot 2023-01-18 at 09 56 35](https://user-images.githubusercontent.com/21013447/213127218-1fc8dee7-6486-4f1a-9252-fcf2514a258f.png)

Only 172kB transferred.

Let's move that to Server Components

```tsx
"use client";

import { useEffect, useState } from "react";

import type { IconType } from "@react-icons/all-files";

const NoIcon = () => <i>No Icon</i>;

const Icons = () => {
  const [iconName, setIconName] = useState("FaBeer");
  const [icon, setIcon] = useState<
    { [key in typeof iconName]: IconType } | null
  >(null);

  useEffect(() => {
    import(`@react-icons/all-files/fa/${iconName}.js`)
      .then((mod) => setIcon(mod))
      .catch((e) => {/* silence the error */});
  }, [iconName]);

  if (!icon) return null;

  const Icon = icon[iconName] || NoIcon;

  return (
    <div>
      <input
        type="text"
        onChange={(e) => setIconName(e.target.value)}
        value={iconName}
      />

      <p>Icon:</p>

      <p>
        <Icon />
      </p>
    </div>
  );
};

export default Icons;
```

And we get this nice result:

https://user-images.githubusercontent.com/21013447/213130219-143acd00-375e-4552-a9d8-afc6626ba274.mov

Each icon is about 1.5kB ~ 


### Unhacking the hack

If you could produce a list of all icons that will be used, then you could make a map that goes `iconName -> lazy import`, so that these are statically known at build time, and then you'd select from that map as needed.

> Btw I choose beer because that's the example icon I found on their page.

**Reply by cereallarceny** [reply]
Wow @icyJoseph , what a fantastic answer. Above and beyond!

So, what I have on my end is a CMS that writes to files which are committed to Github. In fact, the entire CMS is file based. All dynamic icons are selected inside of the CMS and the only time this dynamic icons component is used is in conjunction with the CMS. Therefore, all icons names that we will ever use, must be defined in these CMS files.

When we go to `next build` the application, we should (in theory) be able to pick out which icons are used in the CMS and bundle only those icons, as opposed to bundling one for every single file that exists. You mentioned at the end of your hack solution that this could be theoretically possible if we know all the icons used at build time, which we do.

Where would we create that map and how would we alter the hack solution you proposed to build only the list of icons used in the CMS?

**Reply by cereallarceny** [reply]
To that end - I almost feel like what we're doing here may be a "fool's errand". Surely there's gotta be a better solution here. What I need is:

1. To allow content managers to select icons from a list of all possible icons, which are then stored as a string in some file (i.e. `"FaBeer"`)
2. Be able to reference specific icons from this icon library to be used by the app directory (i.e. `import { FaBeer } from 'react-icons/fa';`

Beyond that, however we want to include those custom icons is totally up to us as developers. I haven't been able to think of a more holistic approach than what you and I are discussing, but if you feel there's a more elegant solution: I'm all ears. 😄

**Reply by cereallarceny** [reply]
One other note: we don't need to use `react-icons`. The reality is that we only use FontAwesome in the first place. I suppose it's not outside the realm of possibility to just integrate directly with that? I don't know, I'm open to any solution here. Thank you again for all your help thus far.

**Reply by cereallarceny** [reply]
Any ideas @icyJoseph ?

**Reply by getanwar** [reply]
Months later I am here to discuss the same issue! I have created a CMS where my users can select from any of the thousands of FontAwesome icons. I don't want to import the entire FA library for every user, rather I want to import only icons necessary for a user. I also thought to implement the same hack @icyJoseph have given here, but at the same time I knew that it's not allowed to write template string or variables inside `import()`. Yet, I wanted to try the hack only to fail. Has anyone found out any alternative solution that works with next.js or any other framework? Frameworks' not important here, using FA icons that ain't known before runtime is!

![image](https://github.com/vercel/next.js/assets/6163268/e21f5470-e731-4e5e-bd99-237537da183f)
